contrib/win32/win32compat/fileio.c

```diff
@@ -1106,39 +1044,159 @@ fileio_readlink(const char *path, char *buf, size_t bufsiz)
+		errno = errno_from_Win32LastError();
 		goto cleanup;
 	}
 
-	if ((output = utf16_to_utf8(resolved_link)) == NULL) {
-		errno = ENOMEM;
+	/* ensure file is actually symlink */
+	if (reparse_buffer->ReparseTag != IO_REPARSE_TAG_SYMLINK) {
+		errno = EINVAL;
+		goto cleanup;
+	}
+
+	/* the symlink structure has a 'Print Name' value that is displayed to the
+	 * user which is different from the actual value it uses for redirection
+	 * called the 'Substitute Name'; since the Substitute Name has an odd format
+	 * that begins with \??\ and it appears that CreateSymbolicLink() always
+	 * formats the PrintName value consistently we will just use that
+	 */
+	int symlink_nonnull_size = reparse_buffer->PrintNameLength;
+	wchar_t * symlink_nonnull = &reparse_buffer->PathBuffer[reparse_buffer->PrintNameOffset / sizeof(WCHAR)];
+
+	/* allocate area to hold a null terminated version of the string */
+	if ((linkpath = malloc(symlink_nonnull_size + sizeof(wchar_t))) == NULL) {
 		goto cleanup;
 	}
 
-	/* ensure output buffer is large enough forward slash and the string */
-	ssize_t out_size = strlen(output);
-	if (1 + out_size > bufsiz) {
-		errno = ENAMETOOLONG;
+	/* copy the data out of the reparse buffer and add null terminator */
+	memcpy(linkpath, symlink_nonnull, symlink_nonnull_size);
+	linkpath[symlink_nonnull_size / sizeof(wchar_t)] = L'\0';
+
+	/* convert link path to utf8 */
+	if ((output = utf16_to_utf8(linkpath)) == NULL) {
+		errno = ENOMEM;
 		goto cleanup;
 	}
 
-	/* copy to output buffer in the forward-slash format: /C:/Path/Target */
+	/* determine if we need to prepend a forward slash to make this look like
+	 * an absolute path C:\Path\Target --> /C:/Path/Target
+	 */
+	int abs_chars = is_absolute_path(output) ? 1 : 0;
+	if (abs_chars)
+		buf[0] = '/';
+
+	/* copy link data to output buffer; per specification, truncation is okay */
 	convertToForwardslash(output);
-	buf[0] = '/';
-	memcpy(buf + 1, output, out_size);
-	ret = out_size + 1;
+	size_t out_size = strlen(output);
+	memcpy(buf + abs_chars, output, min(out_size, bufsiz - abs_chars));
+	ret = (ssize_t) min(out_size + abs_chars, bufsiz);
 
 cleanup:
 
+	if (reparse_buffer)
+		free(reparse_buffer);
+	if (handle != INVALID_HANDLE_VALUE)
+		CloseHandle(handle);
 	if (wpath)
 		free(wpath);
 	if (output)
 		free(output);
-	if (resolved_link)
-		free(resolved_link);
 
-	return (ssize_t) ret;
+	return (ssize_t)ret;
+}
+
+int
+fileio_symlink(const char *target, const char *linkpath)
+{
+	if (target == NULL || linkpath == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	DWORD ret = 0;
+	wchar_t *target_utf16 = utf8_to_utf16(resolved_path(target));
+	wchar_t *linkpath_utf16 = utf8_to_utf16(resolved_path(linkpath));
+	wchar_t *resolved_utf16 = _wcsdup(target_utf16);
+	if (target_utf16 == NULL || linkpath_utf16 == NULL || resolved_utf16 == NULL) {
+		errno = ENOMEM;
+		ret = -1;
+		goto cleanup;
+	}
+
+	/* Relative targets are relative to the link and not our current directory
```

Comment:
I think we are deviating from symlink UNIX implementation?
It didn't talk about target being relative but linkpath being relative.

As per [symlink man page](http://man7.org/linux/man-pages/man2/symlink.2.html), linkpath is relative to the current working directory of the calling process.

![image](https://user-images.githubusercontent.com/23668037/37744198-0020ddec-2d2b-11e8-82e2-7636026c594e.png)

Reply:
Got it. Couple more questions,
1) Line 1160 - if we didn't find "\\" then we should throw an error?
2) Line 1185, 1186 - Don't we need to pass "resolved_utf16" instead of target_utf16 to CreateSymbolicLinkW()?

Reply:
1. No, this is a valid case.  All we're really trying to do is determine where the real target of the symbolic link will end up pointing to so we can determine whether it’s a file or a directory.  If the target is relative, then we compute the target by combining the parent directory of the linkpath with the relative park of target.  At this point in the code linkpath could just be the name of the link or it could be a path with a link name at the end.  In the first case, the target will end up being relative to our current directory so we can just use the target path (line 1160).  If the linkpath is not just the name, it will have one or more directories and then the link name so we trim off the linkname and append the relative target path (line 1161).  The ONLY reason we have do to any of this is because we need to know whether the target is a directory or a file since CreateSymbolicLink() needs to know (unlike Linux).  
2. No, as described above, the resolved_utf16 is only needed to determine if the target or a directory.

Reply:
1. If linkpath_utf16 is just a name (i.e. doesn't have a backslash), then we know that the user wishes to create a link in the current directory that they are running symlink() in.  In this case, the intended target of that link will also be relative to the current directory.  That's why we just use the value of the target.  For example I'm in /tmp and I make a link 'mylink' and point it to 'mydir/mytarget' then path relative to the current directory is simply 'mydir/mytarget' (the operating system will resolve this to /tmp/mydir/mytarget').

2.  No, the '../' is relative to the directory that the link lives in, not the current directory.  So I make '/var/log/link' and point it to '../' it will always resolve to /var/log/.. (aka /var/) regardless of my current directory.  Link resolution has nothing to with your current directory.